src/gradients.jl

```diff
@@ -130,7 +130,12 @@ function finite_difference_gradient(
         end
     end
     cache = GradientCache(df, x, fdtype, returntype, inplace)
-    finite_difference_gradient!(df, f, x, cache, relstep=relstep, absstep=absstep, dir=dir)
+    if typeof(x) <: AbstractArray && fdtype == Val(:central)
```

Comment:
why the check for AbstractArray? I would think that the non-! call should always do a non-mutating version by default.

Reply:
The check for `AbstractArray` is only because I changed the vector->scalar code and not the scalar->vector code.
Since by #157 , the latter should be replaced with jacobian calls.
Once that is done, `x` should be restrained to `Num`.